### PR TITLE
[Flang][OpenMP] Support lowering of simd reductions

### PR DESCRIPTION
**Pull request by skatrak**
This patch enables lowering to MLIR of the reduction clause of `simd` constructs. Lowering from MLIR to LLVM IR remains unimplemented, so at that stage it will result in errors being emitted rather than silently ignoring it as it is currently done.

On composite `do simd` constructs, this lowering error will remain untriggered, as the `omp.simd` operation in that case is currently ignored. The MLIR representation, however, will now contain `reduction` information.

### CODE REVIEW

**Review by tblah — COMMENTED** [review]
The changes look okay, but please either add a proper TODO for the LLVM lowering or don't merge this until after the LLVM lowering is merged.

File: flang/test/Lower/OpenMP/simd.f90
Comment (skatrak) [reply flang/test/Lower/OpenMP/simd.f90:279]: Done

**Comment by skatrak** [comment]
> The changes look okay, but please either add a proper TODO for the LLVM lowering or don't merge this until after the LLVM lowering is merged.

Thanks for the review! There was already a check there, but I updated it to trigger different errors based on the unsupported clause found. Hopefully this addresses your concerns.

File: flang/test/Lower/OpenMP/simd.f90
Comment (tblah) [inline flang/test/Lower/OpenMP/simd.f90:292]: Not this PR, but reading the unredacted IR locally it looks like the loop iteration variable was not privatized.

As I understand it, OpenMP 5.2 section 5.1.1 line 16 seems to say that it should be implicitly firstprivate.

Did I miss something here? If not I'll create an issue.

**Review by tblah — APPROVED** [review]
Thanks for the update. That isn't the normal "not yet implemented" message generated by the TODO macro but I think the error message now looks clean and easy enough to understand:
```
error: loc(...): reduction clause not yet supported
error: loc(...): LLVM Translation failed for operation: omp.simd
error: failed to create the LLVM module
```

File: flang/test/Lower/OpenMP/simd.f90
Comment (skatrak) [reply flang/test/Lower/OpenMP/simd.f90:292]: In this case, I think the previous statement is the one that applies, since it has a single associated loop. So I guess it should actually be linear with a linear-step equal to the increment of the loop. But I believe you're right, since neither that nor lastprivatization in the case of multiple associated loops seem to be currently implemented.

**Comment by skatrak** [comment]
> Thanks for the update. That isn't the normal "not yet implemented" message generated by the TODO macro but I think the error message now looks clean and easy enough to understand

Yes, we're triggering different errors when something is not yet supported in Flang lowering vs. in omp dialect translation to LLVM IR. Should we create an equivalent TODO macro for use there too?

**Comment by kiranchandramohan** [comment]
> Thanks for the update. That isn't the normal "not yet implemented" message generated by the TODO macro but I think the error message now looks clean and easy enough to understand:

Does the Driver report this as a crash?


> > Thanks for the update. That isn't the normal "not yet implemented" message generated by the TODO macro but I think the error message now looks clean and easy enough to understand
> 
> Yes, we're triggering different errors when something is not yet supported in Flang lowering vs. in omp dialect translation to LLVM IR. Should we create an equivalent TODO macro for use there too?

Why don't we issue the TODO in flang lowering? This is best and uniform from the Flang point of view. We can remove the TODO when translation support is added.

**Comment by tblah** [comment]
> Does the Driver report this as a crash?

The driver reports an error exit value but does not print a stack trace.

**Comment by kiranchandramohan** [comment]
If it is not a crash it is probably OK.

If you do not want to change for already implemented lowering, it is fine. But going forward we should try to error out in Flang vs translation.

**Comment by skatrak** [comment]
> If you do not want to change for already implemented lowering, it is fine. But going forward we should try to error out in Flang vs translation.

What is the reason for this preference? In my opinion, it has the disadvantage that we wouldn't be able to independently work on Flang lowering before the translation to LLVM IR for that operation/clause is implemented.

I understand that we don't want to silently ignore clauses by lowering them to MLIR and then not taking them into account in the LLVM IR translation pass. But if we emit an error and return early when we find these operations/clauses there, the result from a user perspective would be the same. At the moment the error message is different, but if that's important it wouldn't take much to add a similar TODO macro to the MLIR to LLVM IR translation as well. Either way we should trigger errors there because if another dialect or frontend decides to introduce operations/clauses for which we have no translation to LLVM IR implemented, we should emit errors instead of crashing/ignoring them.

**Comment by kiranchandramohan** [comment]
> > If you do not want to change for already implemented lowering, it is fine. But going forward we should try to error out in Flang vs translation.
> 
> What is the reason for this preference? In my opinion, it has the disadvantage that we wouldn't be able to independently work on Flang lowering before the translation to LLVM IR for that operation/clause is implemented.

This is to present a uniform way in the flang compiler to message to the user that a feature is not implemented. As you know the user does not care about which part of the compiler does not support the feature. Lowering is currently the place where this is implemented for all features including OpenMP features and we have full control over this.

> I understand that we don't want to silently ignore clauses by lowering them to MLIR and then not taking them into account in the LLVM IR translation pass. But if we emit an error and return early when we find these operations/clauses there, the result from a user perspective would be the same. At the moment the error message is different, but if that's important it wouldn't take much to add a similar TODO macro to the MLIR to LLVM IR translation as well. Either way we should trigger errors there because if another dialect or frontend decides to introduce operations/clauses for which we have no translation to LLVM IR implemented, we should emit errors instead of crashing/ignoring them.

Yes, we should provide appropriate errors in translation if a feature/clause/directive/option is not supported. Adding a similar TODO macro to the MLIR to LLVM IR translation would also work. But I am not sure whether it will be accepted since it is not the way things are done in MLIR. But since at the moment it is only Flang work that uses this it might be OK. Would it be possible to catch this in the call to the translation and issue a TODO in Flang codegen or Driver?


We can see the worst-case issue in this distribute construct translation issue where it presents as a crash.
https://github.com/llvm/llvm-project/issues/109755

**Comment by skatrak** [comment]
> This is to present a uniform way in the flang compiler to message to the user that a feature is not implemented. As you know the user does not care about which part of the compiler does not support the feature. Lowering is currently the place where this is implemented for all features including OpenMP features and we have full control over this.

I'll try to explain a bit better my concern with triggering TODOs in Flang lowering for everything that's not fully supported. Let's say we define the MLIR representation for a given clause. At that point we should be unblocking two lines of work: translation of that clause to LLVM IR, as part of operations that accept it, and Flang lowering to MLIR of that clause and attaching it to the applicable operations. These are two independent features that both need to come together to actually support the new clause by Flang.

However, if we must raise TODO errors in Flang lowering for that clause until it is implemented in both places, then we're artificially creating a dependency. We aren't able to implement and test the lowering to MLIR in Flang because we can't remove the TODO until the other part is also implemented. My thinking is that, by triggering the errors at the point we're actually missing an implementation, we're able to work on both parts independently, since the "contract" between them is the definition of the MLIR operation / clause. From a user perspective, from where we trigger errors doesn't matter, as long as error messages convey the same information.

> Yes, we should provide appropriate errors in translation if a feature/clause/directive/option is not supported. Adding a similar TODO macro to the MLIR to LLVM IR translation would also work. But I am not sure whether it will be accepted since it is not the way things are done in MLIR. But since at the moment it is only Flang work that uses this it might be OK.
> 
> We can see the worst-case issue in this distribute construct translation issue where it presents as a crash. #109755

Looking a bit into this crash, I think the problem here is not only that we introduced an `omp.distribute` operation for which we don't have lowering to LLVM IR implemented yet, but that we're also not checking in the OMPIRBuilder whether the callbacks we're calling succeeded before continuing. I believe the crash would go away if we checked this, and be left with just the error messages ("unsupported OpenMP operation: omp.distribute", "LLVM Translation failed for operation: omp.distribute", ...) and a non-zero exit code. The error propagation is already there in the ModuleTranslation and OpenMPToLLVMIRTranslation, we're just missing some spots in OMPIRBuilder to make sure the compiler doesn't crash when an error is triggered somewhere in the process.

Adding the macro would be mostly to ensure every TODO message is formatted the same way, possibly like Flang lowering, but it's not really necessary if printing the compiler source location of where the error was triggered or defining it as a macro goes against conventions in MLIR.

> Would it be possible to catch this in the call to the translation and issue a TODO in Flang codegen or Driver?

I can't think of a way to do that. We can detect whether translation to LLVM IR failed and then raise an error, but we wouldn't know if it was a TODO that we hit or for what operation it was. I think the functionally equivalent version to that would be to fix the current error handling issues in the translation stage and update error messages to state that it's not yet implemented.

One thing I'd like to propose is to merge this PR so that there is an error triggered for 'simd' but not for 'do simd' reductions (even if it's not a Flang lowering TODO), and work on improving the state of error propagation in the MLIR to LLVM IR translation stage to prevent compiler crashes. This has to be done either way eventually, but it also perhaps helps with concerns of letting Flang lowering produce MLIR that is potentially not yet translatable to LLVM IR.

**Comment by kiranchandramohan** [comment]
> One thing I'd like to propose is to merge this PR so that there is an error triggered for 'simd' but not for 'do simd' reductions (even if it's not a Flang lowering TODO),

+1.

**Review by kiranchandramohan — APPROVED** [review]
LG

**Review by NimishMishra — APPROVED** [review]
LGTM. I'm okay as well with merging this PR.

**Comment by skatrak** [comment]
This PR triggered a Fujistu testsuite regression: [LLVM-1403](https://linaro.atlassian.net/browse/LLVM-1403). I've looked at it and the problem is related to the fact that we currently ignore the inner loop wrapper in 'do simd' during MLIR to LLVM IR translation. I'll create another PR to address this shortly.

**Pull request merged**
